### PR TITLE
Added ability to specify tag names in selectors

### PR DESCRIPTION
**Pull request by avimehta**
Roll forward #4617 with bug fixes and tests.

First commit is same as #4368 
Second commit includes all the bugfixes.
Third commit adds the tests. 

/to @dvoytenko

### CODE REVIEW

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 1]
```diff
@@ -98,7 +100,8 @@ export function isVisibilitySpecValid(config) {
   }
 
   const spec = config['visibilitySpec'];
-  if (!spec['selector'] || spec['selector'][0] != '#') {
+  const selector = spec['selector'];
+  if (!selector || (selector[0] != '#' && selector.indexOf('amp-') != 0)) {
     user().error('Visibility spec requires an id selector');
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Change the message please.

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 2]
```diff
@@ -236,11 +263,21 @@ export class Visibility {
...
    */
-  listenOnce(config, callback, shouldBeVisible) {
-    const element = this.win_.document.getElementById(config['selector']
-        .slice(1));
+  listenOnce(config, callback, shouldBeVisible, analyticsElement) {
+    const selector = config['selector'];
+    const element = getElement(selector, analyticsElement,
+        config['selectionMethod']);
+    if (!element) {
+      user().error('Element not found for visibilitySpec: ' + selector);
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: `user().error()` requires a TAG as a first argument and, also, it doesn't fail by itself, but logs an error asynchronously. So, the code will proceed to the following lines with `getResourceForElement(null)` below.

If it'd work better, you can use `user().assert(element, '....')` which will fail immediately.

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 3]
```diff
@@ -236,11 +263,21 @@ export class Visibility {
...
+    }
     const res = this.resourcesService_.getResourceForElement(element);
+    if (!res) {
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: FYI: `getResourceForElement` fails if it can't find resource, I believe.

**Comment by dvoytenko** [comment]
LGTM with few comments.

**Comment by avimehta** [comment]
ptal. I had to make substantial changes to test to accommodate for using `.assert`. Another look at the changes to make sure I am not doing anything stupid.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 3 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: oh. I added a try catch block to catch that and show a more meaningful error. Hope that's okay.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 2 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: switched to assert

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 1 above)
Comment (avimehta) [inline extensions/amp-analytics/0.1/visibility-impl.js]: done.

**Comment by avimehta** [comment]
@dvoytenko I have been seeing these tests fail on some of the runs. ANy idea why this is happening? The tests pass if I rerun them. https://travis-ci.org/ampproject/amphtml/builds/154618433

**Comment by dvoytenko** [comment]
@avimehta Test failures appear to be unrelated. But please ping Erwin if you keep seeing them.

**Comment by dvoytenko** [comment]
LGTM

**Pull request merged**